File: dawn_haskell/src/Xfast_trie.hs
module XFastTrie
  (XFastTrie,
    create,
    insert,
    predecessor
  ) where

import Data.Bits
import Data.Array.IO

data XFastTrie = XFastTrie
  { bits :: IOArray Int Bool
  , ones :: [Int]
  , size :: Int
  }

create :: Int -> IO XFastTrie
create w = do
  let u = 2 ^ w
  bits <- newArray (0, 2 * u - 2) False
  return $ XFastTrie bits [] 0

insert :: XFastTrie -> Int -> IO XFastTrie
insert (XFastTrie bits ones size) x =
  insertAux bits ones size x 0
 where
  insertAux bits ones size x i
    | i >= sizeOfIOArray bits = return $ XFastTrie bits ones (size + 1)
    | otherwise = do
      let bit = x .&. (1 `shiftL` i) /= 0
      let index = 1 `shiftL` (sizeOfIOArray bits - 1 - i)
      writeArray bits (index - 1) (bit || bits `indexArray` (index - 1))
      if bit
        then insertAux bits (index : ones) size x (i + 1)
        else insertAux bits ones size x (i + 1)

predecessor :: XFastTrie -> Int -> IO (Maybe Int)
predecessor (XFastTrie bits ones _) x = predAux bits ones x 0
 where
  predAux bits ones x i
    | i >= sizeOfIOArray bits = return Nothing
    | otherwise = do
      let index = 1 `shiftL` (sizeOfIOArray bits - 1 - i)
      if bits `indexArray` (index - 1)
        then do
          let leftChild = 2 * index
          let rightChild = leftChild + 1
          if bits `indexArray` (rightChild - 1)
            then predAux bits ones x (i + 1)
            else return $ Just $ head ones
        else predAux bits ones x (i + 1)

File: dawn_rust/src/xfast_trie.rs
use std::collections::VecDeque;

pub struct XFastTrie {
    bits: Vec<bool>,
    ones: VecDeque<usize>,
    size: usize,
}

impl XFastTrie {
    pub fn create(w: u32) -> Self {
        let u = 1 << w;
        let bits = vec![false; 2 * u as usize - 1];
        XFastTrie {
            bits,
            ones: VecDeque::new(),
            size: 0,
        }
    }

    pub fn insert(&mut self, x: usize) {
        self.insert_aux(x, 0);
    }

    fn insert_aux(&mut self, x: usize, i: usize) {
        if i >= self.bits.len() {
            self.size += 1;
            return;
        }
        let bit = (x & (1 << i)) != 0;
        let index = 1 << (self.bits.len() - 1 - i);
        self.bits[index - 1] = self.bits[index - 1] || bit;
        if bit {
            self.ones.push_front(index);
            self.insert_aux(x, i + 1);
        } else {
            self.insert_aux(x, i + 1);
        }
    }

    pub fn predecessor(&self, x: usize) -> Option<usize> {
        self.predecessor_aux(x, 0)
    }

    fn predecessor_aux(&self, x: usize, i: usize) -> Option<usize> {
        if i >= self.bits.len() {
            return None;
        }
        let index = 1 << (self.bits.len() - 1 - i);
        if self.bits[index - 1] {
            let left_child = 2 * index;
            let right_child = left_child + 1;
            if self.bits[right_child - 1] {
                self.predecessor_aux(x, i + 1)
            } else {
                self.ones.front().copied()
            }
        } else {
            self.predecessor_aux(x, i + 1)
        }
    }
}